Feature/#1369 add rating to text exercise

JonasPetry opened the pull request:
<!-- Thanks for contributing to Artemis! Before you submit your pull request, please make sure to check the following boxes by putting an x in the [ ] (don't: [x ], [ x], do: [x]) -->
<!-- If your pull request is not ready for review yet, create a draft pull request! -->

### Checklist
- [X] I tested *all* changes and *all* related features with different users (student, tutor, instructor, admin) on the test server https://artemistest.ase.in.tum.de.
- [x] Server: I added multiple integration tests (Spring) related to the features
- [X] Server: I added `@PreAuthorize` and check the course groups for all new REST Calls (security)
- [X] Server: I implemented the changes with a good performance and prevented too many database calls
- [X] Server: I documented the Java code using JavaDoc style.
- [X] Client: I documented the TypeScript code using JSDoc style.
- [X] Client: I added multiple integration tests (Jest) related to the features
- [X] Client: I added multiple screenshots/screencasts of my UI changes
- [X] Client: I translated all the newly inserted strings into German and English

### Motivation and Context
This feature allows students to rate the feedback they got on a result level in text exercises.

### Description
- Created a RatingModule that can be used for every exercise 
- Added a REST route
- Inserted the newly created RatingComponent into the text exercises and styled it

### Steps for Testing
<!-- Please describe in detail how the reviewer can test your changes. -->

1. Log in to Artemis
2. Navigate to a text-exercise that has feedback
2.1 If such does not exist do the following:
2.2 Create a new Text exercise
2.3 Submit a solution for it
2.4 Assess this text exercise
3. Navigate to "view result" on that text exercise
4. See 5 empty stars at the bottom of the page
5. Rate the result
6. Update the rating by giving it another amount of stars
6. Reload the page and see that the rating was saved

### Screenshots
<!-- Add screenshots to demonstrate the changes in the UI. -->
<!-- Create a GIF file from a screen recording in a docker container https://toub.es/2017/09/11/high-quality-gif-with-ffmpeg-and-docker/ -->
<img width="1440" alt="image" src="https://user-images.githubusercontent.com/51994246/82139073-a5bc1480-9825-11ea-9a41-7eb902f2c5d3.png">

krusche commented:
Can you please implement this feature independent of the exercise?
It might be cumbersome for students to rate all feedbacks individually. There can easily be 15-20 feedback elements, so I suggest to have this rating on the result level

JonasPetry commented:
> Can you please implement this feature independent of the exercise?
> It might be cumbersome for students to rate all feedbacks individually. There can easily be 15-20 feedback elements, so I suggest to have this rating on the result level

@krusche @jpbernius and I wanted to start using it with the text exercise in order to see whether someone uses it at all. Once we have some insights I would make it available to every exercise type.

Currently we are assessesing every feedback individually because we believe that students are most likely to rate when they are not happy with the feedback. Being this granular would allow to exclude certain feedbacks from the automatic-suggestion based on bad student rating. I definetly get your point that students might not use it for every feedback!

krusche commented:
> > Can you please implement this feature independent of the exercise?
> > It might be cumbersome for students to rate all feedbacks individually. There can easily be 15-20 feedback elements, so I suggest to have this rating on the result level
> 
> @krusche @jpbernius and I wanted to start using it with the text exercise in order to see whether someone uses it at all. Once we have some insights I would make it available to every exercise type.
> 
> Currently we are assessesing every feedback individually because we believe that students are most likely to rate when they are not happy with the feedback. Being this granular would allow to exclude certain feedbacks from the automatic-suggestion based on bad student rating. I definetly get your point that students might not use it for every feedback!

**I don't agree!**

1) we can't just implement such features for one exercise type, we have to make sure the feature makes sense for all exercise types! If exercises with common workflows (such as modeling, text, file upload) deviate too much, the user experience and the development become inconsistent, therefore I would ask you to implement this exercise independent in a common Angular component that can be reused across all exercise types

2) Adding ratings on the feedback level does not make sense in my opinion! When you buy something on Amazon, do you rate the whole item or do you rate every single component of it individually? If we want to get more details on the rating of one result, we can still ask students to add additional text to explain the rating in a second step.

JonasPetry commented:
> > > Can you please implement this feature independent of the exercise?
> > > It might be cumbersome for students to rate all feedbacks individually. There can easily be 15-20 feedback elements, so I suggest to have this rating on the result level
> > 
> > 
> > @krusche @jpbernius and I wanted to start using it with the text exercise in order to see whether someone uses it at all. Once we have some insights I would make it available to every exercise type.
> > Currently we are assessesing every feedback individually because we believe that students are most likely to rate when they are not happy with the feedback. Being this granular would allow to exclude certain feedbacks from the automatic-suggestion based on bad student rating. I definetly get your point that students might not use it for every feedback!
> 
> **I don't agree!**
> 
>     1. we can't just implement such features for one exercise type, we have to make sure the feature makes sense for all exercise types! If exercises with common workflows (such as modeling, text, file upload) deviate too much, the user experience and the development become inconsistent, therefore I would ask you to implement this exercise independent in a common Angular component that can be reused across all exercise types
> 
>     2. Adding ratings on the feedback level does not make sense in my opinion! When you buy something on Amazon, do you rate the whole item or do you rate every single component of it individually? If we want to get more details on the rating of one result, we can still ask students to add additional text to explain the rating in a second step.

Ok

jpbernius commented:
> 1. we can't just implement such features for one exercise type, we have to make sure the feature makes sense for all exercise types! If exercises with common workflows (such as modeling, text, file upload) deviate too much, the user experience and the development become inconsistent, therefore I would ask you to implement this exercise independent in a common Angular component that can be reused across all exercise types

This is the plan indeed. It should be a component that can be used for all exercise types.
In the first phase however, we would integrate it for text exercises and only add it later to other exercise types.

> 2. Adding ratings on the feedback level does not make sense in my opinion! When you buy something on Amazon, do you rate the whole item or do you rate every single component of it individually?

If you place an order on amazon for 3 products, your rate every product individually.
I think we talked about this in the past and derived at a different conclusion.
Having it at a granular level has a few advantages:
1. We can easily see how automatic assessments perform compared to adapted or manual assessments.
2. We can use this to identify a single grading criteria with an unhelpful (or especially helpful) feedback text.
3. .. identify tests in programming exercises with good or bad automatic feedback.

In the contrary, I think having this on result level does not provide us with much information other than a vague idea how students receive their feedback. Not giving us any benefit from the structured feedback approach we have in modeling, text and also programming exercises.

>If we want to get more details on the rating of one result, we can still ask students to add additional text to explain the rating in a second step.

I do not think this is a practical approach, because this cannot be automated, making this just another feedback box nobody is ever going to read.

filip-gregurevic commented on src/main/webapp/app/exercises/shared/rating/rating.service.spec.ts:
```diff
@@ -0,0 +1,16 @@
+    });
+
+    it('should be created', () => {
+        expect(service).toBeTruthy();
```
nice test case 💯

filip-gregurevic commented on src/main/webapp/app/exercises/shared/rating/rating.component.spec.ts:
```diff
@@ -0,0 +1,24 @@
+    });
+
+    it('should create', () => {
+        expect(component).toBeTruthy();
```
👍

krusche commented on src/main/java/de/tum/in/www1/artemis/domain/Rating.java:
```diff
@@ -0,0 +1,71 @@
+    private Integer rating;
+
+    @MapsId
+    @OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
```
I don’t think cascade makes sense here. This would mean a result is deleted when a rating is deleted

JonasPetry replied:
From my understanding Cascade.ALL propagates all operations — including Hibernate-specific ones — from a parent to a child entity. In an uni-directional 1-1 mapping using "MapsId", the referenced table is the parent entity and the referencing one is the child entity. Therefore I would delete the rating if the result ist deleted.

JonasPetry replied:
Nevermind, I've tested it and you are right. I removed the cascading now!

krusche commented on src/main/java/de/tum/in/www1/artemis/service/RatingService.java:
```diff
@@ -0,0 +1,67 @@
+     * @param rating - Rating that should be persisted
+     * @return persisted Rating
+     */
+    @Transactional
```
Please remove transactional here

JonasPetry replied:
I've tried it in a few ways but I honestly did not figure out how to do it without using transactional. I've come across the book "Spring Boot Persistence Best Practices: Optimize Java Persistence" which recommended the way I've implemented it now.

<img width="1552" alt="image" src="https://user-images.githubusercontent.com/51994246/82840775-0c9b9680-9ed4-11ea-90c9-6028aa2b57e6.png">

krusche replied:
The book does not matter, we have design and coding guidelines in Artemis. Please do not use transactional except you have a very good reason for it. I don’t see a good reason in the cases of this PR. Make sure to fetch objects (with related objects) directly from the database to avoid additional fetches. There are plenty of examples in the repositories how to do that

JonasPetry replied:
I think the `@Transactional` is needed in this case, because the Id is not automatically set otherwise.

krusche commented on src/main/java/de/tum/in/www1/artemis/service/RatingService.java:
```diff
@@ -0,0 +1,67 @@
+     * @param rating - Updated Rating that should be persisted
+     * @return updated rating
+     */
+    @Transactional
```
Please remove transactional here

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,133 @@
+    private String applicationName;
+
+    @Value("${artemis.continuous-integration.artemis-authentication-token-value}")
+    private String ARTEMIS_AUTHENTICATION_TOKEN_VALUE = "";
```
Do you really need this?

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,133 @@
+     * @param resultId - Id of result that is referenced with the rating
+     * @return Rating or 404 Not Found
+     */
+    @GetMapping("/rating/result/{resultId}")
```
The URL is incorrect. I suggest /results/{resultId}/rating

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,133 @@
+     * @throws URISyntaxException if the Location URI syntax is incorrect
+     */
+    @PostMapping("/rating")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
```
Wrong url

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,133 @@
+     * @return updated Rating
+     */
+    @PutMapping("/rating")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
```
Wrong url

madwau commented on src/main/webapp/app/exercises/shared/rating/rating.service.ts:
```diff
@@ -0,0 +1,38 @@
+    providedIn: 'root',
+})
+export class RatingService {
+    private ratingResourceUrl = SERVER_API_URL + 'api/results/';
```
You're adding the slash twice, resulting in a 500.

![OnPaste 20200526-104359](https://user-images.githubusercontent.com/7109225/82879846-d4c93900-9f3d-11ea-8d1f-ddc831434c8d.png)

madwau commented on src/main/webapp/app/exercises/shared/rating/rating.service.ts:
```diff
@@ -0,0 +1,38 @@
+     * @param rating - Rating for the result
+     */
+    updateRating(rating: Rating): Observable<Rating> {
+        return this.http.put<Rating>(this.ratingResourceUrl + `${rating.id}/rating/${rating.rating}`, rating);
```
All the urls seem inconsistent with the server endpoints. Does `rating.service.ts` still need to be adapted to the latest changes in `RatingResource`?

madwau replied:
The result id needs to be passed in. `${rating.id}` should be `${result.id}`.

madwau replied:
Although I guess it's technically correct since they are the same when `MapsId` is used.

madwau replied:
I think the only real issue is in `setRating` then. It fails since `rating.id` is still undefined at that point.

madwau commented on src/main/java/de/tum/in/www1/artemis/domain/Rating.java:
```diff
@@ -0,0 +1,71 @@
+
+    @Override
+    public String toString() {
+        return "Feedback{" + "id=" + getId() + ", rating='" + getRating() + "'" + ", resultId='" + getId() + "'}";
```
```suggestion
        return "Rating{" + "id=" + getId() + ", rating='" + getRating() + "'}";
```

madwau commented on src/main/java/de/tum/in/www1/artemis/service/RatingService.java:
```diff
@@ -0,0 +1,71 @@
+     * @return updated rating
+     */
+    public Rating updateRating(Long resultId, Integer ratingValue) {
+        Rating update = this.ratingRepository.findById(resultId).orElseThrow();
```
Maybe better call it `updatedRating`?

madwau replied:
`this.` is not needed.

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,120 @@
+    }
+
+    /**
+     * Return Rating referencing resultId or 404 Not Found
```
I think we should avoid raising a 404 simply because a result doesn't have a rating yet. I think a 404 should mostly be reserved for actual unexpected not found errors since it will also end up in Sentry and the error statistics. Having a 404 raised every time someone looks at an assessed submission without result rating would generate a lot of 404s.

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,120 @@
+    @GetMapping("/results/{resultId}/rating")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Rating> getRatingForResult(@PathVariable Long resultId) {
+        Optional<Rating> rating = this.ratingService.findRatingByResultId(resultId);
```
`this.` is not needed.

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,120 @@
+            return forbidden();
+        }
+
+        Rating result = this.ratingService.saveRating(resultId, ratingValue);
```
`this.` is not needed.

madwau replied:
The name `result` is pretty confusing here since it's actually a `Rating` and not a `Result`. Better call it `rating` or `savedRating`.

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,120 @@
+    /**
+     * Persist a new Rating
+     *
+     * @param resultId    - Id of the rating that should be updated
```
Can be confusing for other developers since it assumes knowledge of the shared primary key.

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,120 @@
+    /**
+     * Update a Rating
+     *
+     * @param resultId    - Id of the rating that should be updated
```
Can be confusing for other developers since it assumes knowledge of the shared primary key.

madwau commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,120 @@
+            return forbidden();
+        }
+
+        Rating result = this.ratingService.updateRating(resultId, ratingValue);
```
`this.` is not needed.

madwau replied:
The name `result` is pretty confusing here since it's actually a `Rating` and not a `Result`. Better call it `rating` or `updatedRating`.

madwau commented on src/main/webapp/app/exercises/shared/rating/rating.component.ts:
```diff
@@ -0,0 +1,59 @@
+        );
+    }
+
+    constructor(public ratingService: RatingService) {}
```
Constructor should be at the top.

madwau commented on src/main/webapp/app/exercises/shared/rating/rating.component.ts:
```diff
@@ -0,0 +1,59 @@
+export class RatingComponent {
+    // public ratingValue = 2;
+    public rating: Rating;
+    private _result: Result;
```
I don't think we use this syntax of prefixing private attributes with an underscore at the moment.

madwau commented on src/main/webapp/app/exercises/shared/rating/rating.component.ts:
```diff
@@ -0,0 +1,59 @@
+    styleUrls: ['./rating.component.scss'],
+})
+export class RatingComponent {
+    // public ratingValue = 2;
```
Can this be removed?

madwau commented on src/main/webapp/app/exercises/shared/rating/rating.service.ts:
```diff
@@ -0,0 +1,38 @@
+    constructor(private http: HttpClient) {}
+
+    /**
+     * Update the student rating for feedback on the server.
```
I think the word `update` should be avoided here since it could be misinterpreted as: The rating already exists and we want to update it (even though this is not the case here, it is being created).

madwau commented on src/main/webapp/app/exercises/shared/rating/rating.service.ts:
```diff
@@ -0,0 +1,38 @@
+     * Update the student rating for feedback on the server.
+     * @param rating - Rating for the result
+     */
+    setRating(rating: Rating): Observable<Rating> {
```
I think `createRating` might be better to avoid confusion with `updateRating`.

madwau commented on src/main/webapp/app/exercises/shared/rating/rating.component.ts:
```diff
@@ -0,0 +1,59 @@
+     * @param result
+     */
+    @Input()
+    public set result(result: Result) {
```
Is this construction needed compared to a normal `@Input` declaration and loading the result in `ngOnInit`?

madwau commented on src/test/java/de/tum/in/www1/artemis/RatingResourceIntegrationTest.java:
```diff
@@ -0,0 +1,136 @@
+    @Test
+    @WithMockUser(value = "student1", roles = "USER")
+    public void testCreateRating_asUser() throws Exception {
+        request.post("/api/results/" + result.getId() + "/rating/" + rating.getRating(), null, HttpStatus.CREATED);
```
It would be good to add some assertions on the saved entity, e.g. that it belongs to the correct result and that the rating value is correct.

madwau commented on src/test/java/de/tum/in/www1/artemis/RatingResourceIntegrationTest.java:
```diff
@@ -0,0 +1,136 @@
+    @Test
+    @WithMockUser(value = "student1", roles = "USER")
+    public void testUpdateRating_asUser() throws Exception {
+        Rating result = ratingService.saveRating(this.result.getId(), rating.getRating());
```
The name `result` is confusing in this context. Should be changed to `rating` or `savedRating`. Then the `this.` is also not needed anymore.

madwau commented on src/test/java/de/tum/in/www1/artemis/RatingResourceIntegrationTest.java:
```diff
@@ -0,0 +1,136 @@
+    @WithMockUser(value = "student1", roles = "USER")
+    public void testUpdateRating_asUser() throws Exception {
+        Rating result = ratingService.saveRating(this.result.getId(), rating.getRating());
+        request.put("/api/results/" + result.getId() + "/rating/" + rating.getRating(), null, HttpStatus.OK);
```
Assertions are missing. Also, what is being tested here? Wouldn't it be better if an actual change to the rating value was tested and verified?

madwau commented on src/test/java/de/tum/in/www1/artemis/RatingResourceIntegrationTest.java:
```diff
@@ -0,0 +1,136 @@
+    @Test
+    @WithMockUser(value = "tutor1", roles = "TA")
+    public void testUpdateRating_asTutor_FORBIDDEN() throws Exception {
+        Rating result = ratingService.saveRating(this.result.getId(), rating.getRating());
```
See above.

madwau commented on src/main/java/de/tum/in/www1/artemis/domain/Rating.java:
```diff
@@ -0,0 +1,71 @@
+    @MapsId
+    @OneToOne(fetch = FetchType.LAZY, cascade = CascadeType.ALL)
+    @JoinColumn(name = "id")
+    private Result result;
```
Depending on the usage a `@JsonIgnore` could be added since right now the whole `result` is always serialized when a rating is returned by the server although it's not actually needed for the current use case as far as I can see.

JonasPetry commented:
> I like the idea of the feature and the UI looks nice!
> 
> I think code-wise it still needs some work here and there. Maybe you can look into my comments.

@madwau I've implemented all the changes you requested. Thanks for reviewing again!


> See my in-line comments, client tests are missing, please add those

@krusche I've added the client test cases and left comments on the two remaining requested changes. I hope we can find a solution for this soon in order to add this feature to the next release 💪

krusche commented:
Please also double check the coverage of the tests. Ideally, they should be 100% for the new and changed server resources and server services and for the client files. If this is not reachable, it must be at least 80%!

JonasPetry commented:
> Please also double check the coverage of the tests. Ideally, they should be 100% for the new and changed server resources and server services and for the client files. If this is not reachable, it must be at least 80%!

@krusche The server resources already had 100% coverage. I've adjusted the tests for the rating.component to cover 100% of the lines. Regarding the service, these methods don't contain any logic. Therefore I suggest not to write tests for a simple, one-line REST call.

I've also removed `@MapsId` according to your feedback.

sascha11110 commented on src/main/java/de/tum/in/www1/artemis/domain/Rating.java:
```diff
@@ -0,0 +1,70 @@
+    @Column(name = "rating")
+    private Integer rating;
+
+    @OneToOne()
```
Can you remove the unnecessary parentheses here?

JonasPetry replied:
Done

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,115 @@
+    @GetMapping("/results/{resultId}/rating")
+    @PreAuthorize("hasAnyRole('USER', 'TA', 'INSTRUCTOR', 'ADMIN')")
+    public ResponseEntity<Optional<Rating>> getRatingForResult(@PathVariable Long resultId) {
+        Optional<Rating> rating = ratingService.findRatingByResultId(resultId);
```
This means, everyone can see every rating in all courses, not a good solution. 
I think we should only allow this to users for results for their own participations.
TAs and instructors can see all ratings in the course in which they are TAs / instructors.
Admins can see all ratings

JonasPetry replied:
I've discussed this with @jpbernius and for now we will only allow the student who rated the result to see it. 

Your suggested changes would follow at a later point. Is that ok for you?

krusche replied:
That would be ok. It's just important to have access control here. You can easily integrate admins as well, I think this would be good! Please leave a TODO for tutors and instructors

JonasPetry replied:
Admins are now allowed as well and a TODO for instructors is added!

krusche commented on src/main/webapp/app/exercises/shared/rating/rating.component.ts:
```diff
@@ -0,0 +1,58 @@
+        if (!this.result || !this.result.submission) {
+            return;
+        }
+        this.ratingService.getRating(this.result.id).subscribe((rating) => {
```
this means we have one additional REST call, when navigating into the page that shows the rating / allows the user to interact with it.

Can you please do the following: 
- Only load the rating if it is not yet available (in the passed result) --> You might need to add rating as reference to result then
- Check if you can load the rating together with the result. (using the fetch mechanism in the repository)

jpbernius replied:
We use the same approach for complaints.
For those, we also do an extra network call to see if they are present.
I think we need to do it that way if we want to keep our design goal of this being independent and integratable by a single line.
Right now the Result does not know about ratings.

If we change this, we would need to update the Resources we well to include this for other exercise types.

krusche replied:
That's exactly my concern, we then have 3 request: result, complaint and rating.
We can keep this for now, but I would like to have 1 request here in the future

krusche commented on src/main/java/de/tum/in/www1/artemis/web/rest/RatingResource.java:
```diff
@@ -0,0 +1,126 @@
+     */
+    private boolean checkIfUserIsOwnerOfSubmission(Long resultId) {
+        User user = userService.getUser();
+        Result res = resultService.findOne(resultId);
```
can you reduce this to one database access to find the participation directly?

JonasPetry replied:
Done

Varnika-Tyagi commented:
Works as expected on TS1
![Rating in text exercises](https://user-images.githubusercontent.com/17163093/83565619-008c8600-a51f-11ea-964a-cf49101956a9.PNG)

florian-glombik commented:
I am too late for a review, sorry about that.

The only thing that I've noticed, is that it is not possible to give 0 stars as soon as any (1+) stars were given. I guess it is intended like that.
